rclpy/src/rclpy/utils.cpp

```diff
@@ -277,14 +277,117 @@ throw_if_unparsed_ros_args(py::list pyargs, const rcl_arguments_t & rcl_args)
-    pyqos_profile = rclpy_common_convert_to_qos_dict(&rcl_action_qos_profile_status_default);
+    pyqos_profile = common_convert_to_qos_dict(&rcl_action_qos_profile_status_default);
   } else {
     std::string error_text = "Requested unknown rmw_qos_profile: ";
     error_text += rmw_profile;
     throw std::runtime_error(error_text);
   }
-  return py::reinterpret_steal<py::dict>(pyqos_profile);
+  return pyqos_profile;
+}
+
+py::object
+_convert_to_py_topic_endpoint_info(const rmw_topic_endpoint_info_t * topic_endpoint_info)
+{
+  py::list py_endpoint_gid = py::list(RMW_GID_STORAGE_SIZE);
+  for (size_t i = 0; i < RMW_GID_STORAGE_SIZE; i++) {
+    py::object py_val_at_index = py::int_(topic_endpoint_info->endpoint_gid[i]);
+    py_endpoint_gid[i] = py_val_at_index;
+  }
+
+  // Create dictionary that represents rmw_topic_endpoint_info_t
+  py::dict py_endpoint_info_dict;
+  // Populate keyword arguments
+  // A success returns 0, and a failure returns -1
+  py_endpoint_info_dict["node_name"] = py::str(topic_endpoint_info->node_name);
+  py_endpoint_info_dict["node_namespace"] = py::str(topic_endpoint_info->node_namespace);
+  py_endpoint_info_dict["topic_type"] = py::str(topic_endpoint_info->topic_type);
+  py_endpoint_info_dict["endpoint_type"] =
+    py::int_(static_cast<int>(topic_endpoint_info->endpoint_type));
+  py_endpoint_info_dict["endpoint_gid"] = py_endpoint_gid;
+  py_endpoint_info_dict["qos_profile"] =
+    common_convert_to_qos_dict(&topic_endpoint_info->qos_profile);
+
+  return py_endpoint_info_dict;
+}
+
+py::list
+convert_to_py_topic_endpoint_info_list(const rmw_topic_endpoint_info_array_t * info_array)
+{
+  if (!info_array) {
+    throw RCLError("rmw_topic_endpoint_info_array_t pointer is empty");
+  }
+
+  py::list py_info_array = py::list(info_array->size);
+
+  for (size_t i = 0; i < info_array->size; ++i) {
+    rmw_topic_endpoint_info_t topic_endpoint_info = info_array->info_array[i];
+    // add this dict to the list
+    py_info_array[i] = _convert_to_py_topic_endpoint_info(&topic_endpoint_info);
+  }
+  return py_info_array;
+}
+
+/**
+ * Mirrors the struct rmw_qos_profile_t from rmw/types.h
+ */
+typedef struct rclpy_qos_profile
+{
+  int depth;
+  int history;
+  int reliability;
+  int durability;
+  py::object lifespan;
+  py::object deadline;
+  int liveliness;
+  py::object liveliness_lease_duration;
+  bool avoid_ros_namespace_conventions;
+} rclpy_qos_profile_t;
+
+static
+py::object
+_convert_rmw_time_to_py_duration(const rmw_time_t * duration)
+{
+  auto pyduration_module = py::module::import("rclpy.duration");
+  py::object pymetaclass = pyduration_module.attr("Duration");
+  // to bring in the `_a` literal
+  using namespace pybind11::literals;  // NOLINT
+  return pymetaclass("seconds"_a = duration->sec, "nanoseconds"_a = duration->nsec);
+}
+
+py::dict
+common_convert_to_qos_dict(const rmw_qos_profile_t * qos_profile)
+{
+  // Convert rmw members to Python objects
+  rclpy_qos_profile_t rclpy_qos;
```

Comment:
I don't think this struct is needed anymore. It looks like it's job was to make cleanup easier, but now the pybind11 type's have destructors. The conversion results can be put into the python dict `pyqos_kwargs` directly.